### PR TITLE
refactor(iroh-bytes): remove handshake and rely on ALPN only for protocol negotiation

### PR DESCRIPTION
**Pull request by dignifiedquire**
Closes the simple part of #964.

### CODE REVIEW

File: iroh-bytes/src/protocol.rs [hunk 1]
```diff
@@ -22,11 +21,14 @@ pub(crate) const MAX_MESSAGE_SIZE: usize = 1024 * 1024 * 100;
...
 pub const VERSION: u64 = 2;
 
-#[derive(Deserialize, Serialize, Debug, PartialEq, Eq, Clone, MaxSize)]
+#[derive(Deserialize, Serialize, Debug, PartialEq, Eq, Clone)]
 pub(crate) struct Handshake {
     pub version: u64,
 }
 
+/// Serialized length of [Handshake].
+pub(crate) const HANDSHAKE_SIZE: usize = 1;
```
Comment (rklaehn) [inline iroh-bytes/src/protocol.rs:30]: That's a nice handshake size. But what will we do after version 255? Maybe we should do a multihandshake crate 🤣 ...

File: iroh-bytes/src/protocol.rs [hunk 2]
```diff
@@ -284,3 +298,17 @@ impl TryFrom<VarInt> for Closed {
...
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_handshake_ser() {
+        let handshake = Handshake::new();
+        let ser = postcard::to_stdvec(&handshake).unwrap();
+        assert_eq!(ser.len(), HANDSHAKE_SIZE);
```
Comment (rklaehn) [inline iroh-bytes/src/protocol.rs:310]: The postcard format is stable, so we could actually assert the byte(s) here as well, right?

**Review by rklaehn — DISMISSED** [review]
Save all the bytes!

File: iroh-bytes/src/protocol.rs (hunk 1 above)
Comment (dignifiedquire) [reply iroh-bytes/src/protocol.rs]: Sorry, I actually ended up increasing the size, this comment made me realize that postcard under the hood still uses varints.. Now we have a fully hard coded handshake, with a lot of protocol versions left 😅

File: iroh-bytes/src/protocol.rs [hunk 3]
```diff
@@ -19,19 +18,8 @@ use crate::util::Hash;
...
-    pub version: u64,
-}
-
-impl Handshake {
-    pub fn new() -> Self {
-        Self { version: VERSION }
-    }
-}
+/// The ALPN used with quic for the iroh bytes protocol.
+pub const ALPN: [u8; 13] = *b"/iroh-bytes/2";
```
Comment (flub) [inline iroh-bytes/src/protocol.rs:22]: you want to lose the `n0/` prefix?

**Review by flub — APPROVED** [review]

File: iroh-bytes/src/get.rs [hunk 4]
```diff
@@ -193,25 +192,15 @@ pub mod get_response_machine {
...
-                write_lp(&mut writer, used).await?;
-            }
-
-            // 2. Send Request
+            // 1. Send Request
             {
                 debug!("sending request");
                 // wrap the get request in a request so we can serialize it
                 let request_bytes = postcard::to_stdvec(&request)?;
                 write_lp(&mut writer, &request_bytes).await?;
```
Comment (rklaehn) [inline iroh-bytes/src/get.rs:200]: at some point we can get rid of this lp as well... But not yet.

**Review by rklaehn — APPROVED** [review]

File: iroh-bytes/src/protocol.rs (hunk 3 above)
Comment (dignifiedquire) [reply iroh-bytes/src/protocol.rs:22]: yeah, seems unnecessary to me,

Comment (flub) [reply iroh-bytes/src/protocol.rs:22]: but why leave the leading `/`?

Comment (dignifiedquire) [reply iroh-bytes/src/protocol.rs:22]: personal preference, not a strong one though

**Pull request merged**